dns_server/src/crawler/mod.rs

```diff
@@ -0,0 +1,560 @@
+            return Err(DnsServerError::Other("Unexpected storage version"));
+        }
+
+        let mut addresses = BTreeMap::new();
+        for address in tx.get_addresses()?.iter().filter_map(|address| address.parse().ok()) {
+            Self::new_address(&mut addresses, address, true);
+        }
+        tx.close();
+
+        if storage_version.is_none() {
+            let mut tx = storage.transaction_rw()?;
+            tx.set_version(STORAGE_VERSION)?;
+            tx.commit()?;
+        }
+
+        // Add addresses that were specified from the command line as reachable
+        for address in config.add_node.iter() {
+            let address = address.parse()?;
+            Self::new_address(&mut addresses, address, true);
+        }
+
+        Ok(Self {
+            config,
+            conn,
+            sync,
+            addresses,
+            peers: BTreeMap::new(),
+            storage,
+            command_tx,
+        })
+    }
+
+    fn handle_conn_request(
+        &mut self,
+        _peer_id: N::PeerId,
+        request_id: N::PeerRequestId,
+        request: PeerManagerRequest,
+    ) {
+        match request {
+            PeerManagerRequest::AddrListRequest(_) => {
+                // Ignored
+            }
+            PeerManagerRequest::AnnounceAddrRequest(AnnounceAddrRequest { address }) => {
+                // TODO: Rate limit AnnounceAddrRequest requests
+                if let Some(address) = TransportAddress::from_peer_address(&address) {
+                    Self::new_address(&mut self.addresses, address, false);
+                }
+            }
+            PeerManagerRequest::PingRequest(PingRequest { nonce }) => {
+                let _ = self.conn.send_response(
+                    request_id,
+                    PeerManagerResponse::PingResponse(PingResponse { nonce }),
+                );
+            }
+        }
+    }
+
+    fn handle_conn_response(
+        &mut self,
+        _peer_id: N::PeerId,
+        _request_id: N::PeerRequestId,
+        _response: PeerManagerResponse,
+    ) {
+        // Ignore all
+    }
+
+    fn handle_outbound_accepted(
+        &mut self,
+        address: N::Address,
+        peer_info: PeerInfo<N::PeerId>,
+        _receiver_address: Option<PeerAddress>,
+    ) {
+        let address_data = self
+            .addresses
+            .get_mut(&address)
+            .expect("address in the Connecting state must be known");
+        let valid = peer_info.network == self.config.network;
+
+        if !valid {
+            log::debug!("invalid peer detected at {}", address.to_string());
+            self.conn.disconnect(peer_info.peer_id).expect("disconnect must succeed");
+            Self::change_address_state(
+                &self.config,
+                &address,
+                address_data,
+                AddressState::Disconnected,
+                &mut self.storage,
+                &self.command_tx,
+            );
+            return;
```

Comment:
Please put both these calls in a separate function. We don't update states within code logic.

Reply:
Sorry, but I don't like this. It seems like extra boilerplate to me. Let's keep it, please.

Reply:
I looked one more time and still think that we should keep the code as it is.
That's because connected peers could only ever be created in the `handle_outbound_accepted` function.
So this new proposed function could only be called from one place - `handle_outbound_accepted`.

Reply:
What if someone in the future needed another way to create/delete peers? Instead of calling this function that does two things, they have to copy two things, hopefully.

Reply:
I added `create_outbound_peer` and `remove_outbound_peer` functions. Should be OK now.
Also renamed `peers` to `outbound_peers` to make it clearer that only outbound peers are stored there.